### ISSUE TITLE
error code 500 when FS sessions high concurrency

### ISSUE THREAD

**Issue by ongnen** [znerol/prometheus-freeswitch-exporter#2]
[10/Oct/2020 09:57:19] "GET /esl?module=default HTTP/1.1" 500 -

**Comment by znerol** [comment]
Hi, thanks for the report. Could you please check whether the connection is denied on the freeswitch side or whether something goes wrong in the exporter.

I suggest to just try to connect using `fs_cli` in such situations (`fs_cli` connects to the esl socket as well). If that fails, then there is likely a problem with freeswitch.

**Comment by znerol** [comment]
Closing this due to inactivity. Please reopen if there is new data available.

**Comment by tyl0r** [comment]
I get this too. I have three machines: two with quite a bit of traffic. It's hit or miss with 500s for the two, but I consistently get a 200 for the one without load. After looking at the network traffic (and your code) it looks like you're looping through each call and uuid_dump'ing them... if a call (channel) goes away then we get "ERR No such channel!" and a 500. 

I have freeswitch_exporter running on prometheus. Am I doing something wrong?

**Comment by znerol** [comment]
> After looking at the network traffic (and your code) it looks like you're looping through each call and uuid_dump'ing them... if a call (channel) goes away then we get "ERR No such channel!" and a 500.

Thanks for this very valuable analysis. I'll gonna check whether there is a better way to handle this situation.

**Comment by znerol** [comment]
Opened PR #4 which potentially fixes that problem. Would you mind checking whether that reduces scrape failures?

**Comment by tyl0r** [comment]
> Opened PR #4 which potentially fixes that problem. Would you mind checking whether that reduces scrape failures?

Sorry for the late reply. I've been away from my computer. I'm not sure if I'm testing this right, but the problem remains. Since the package was already pip installed, I just overwrote collector.py with your edits and cleared the __pycache__

**Comment by znerol** [comment]
Would you mind testing the attached wheel? It contains the fix and also #5 which should give us a better understanding where things are failing.
[prometheus_freeswitch_exporter-1.0.0-py3-none-any.whl.gz](https://github.com/znerol/prometheus-freeswitch-exporter/files/6252382/prometheus_freeswitch_exporter-1.0.0-py3-none-any.whl.gz)

**Comment by tyl0r** [comment]
Here's the output when getting a 500

```
127.0.0.1 - - [06/Apr/2021 08:19:43] "GET /esl?module=default&target=fs1.internal HTTP/1.1" 200 -
127.0.0.1 - - [06/Apr/2021 08:19:48] "GET /esl?module=default&target=fs2.internal HTTP/1.1" 200 -
127.0.0.1 - - [06/Apr/2021 08:19:55] "GET /esl?module=default&target=fs3.internal HTTP/1.1" 200 -
127.0.0.1 - - [06/Apr/2021 08:19:58] "GET /esl?module=default&target=fs1.internal HTTP/1.1" 200 -
Exception thrown while rendering view
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/freeswitch_exporter/http.py", line 103, in view
    return self._views[endpoint](**params)
  File "/usr/local/lib/python3.8/dist-packages/freeswitch_exporter/http.py", line 55, in on_esl
    output = collect_esl(self._config[module], target)
  File "/usr/local/lib/python3.8/dist-packages/freeswitch_exporter/collector.py", line 296, in collect_esl
    return generate_latest(registry)
  File "/usr/local/lib/python3.8/dist-packages/prometheus_client/exposition.py", line 107, in generate_latest
    for metric in registry.collect():
  File "/usr/local/lib/python3.8/dist-packages/prometheus_client/registry.py", line 83, in collect
    for metric in collector.collect():
  File "/usr/local/lib/python3.8/dist-packages/asgiref/sync.py", line 147, in __call__
    return call_result.result()
  File "/usr/lib/python3.8/concurrent/futures/_base.py", line 432, in result
    return self.__get_result()
  File "/usr/lib/python3.8/concurrent/futures/_base.py", line 388, in __get_result
    raise self._exception
  File "/usr/local/lib/python3.8/dist-packages/asgiref/sync.py", line 212, in main_wrap
    result = await self.awaitable(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/freeswitch_exporter/collector.py", line 285, in collect
    await ESLChannelInfo(esl).collect())
  File "/usr/local/lib/python3.8/dist-packages/freeswitch_exporter/collector.py", line 234, in collect
    channelvars = json.loads(result)
  File "/usr/lib/python3.8/json/__init__.py", line 357, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3.8/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.8/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
127.0.0.1 - - [06/Apr/2021 08:20:03] "GET /esl?module=default&target=fs2.internal HTTP/1.1" 500 -
```

**Comment by znerol** [comment]
Great, thanks for that trace. Added an additional check before attempting to deserialize the json (cd1f36c387a9f30bf1a29fb935d50fcdc65393ec). The attached wheel includes that fix.
[prometheus_freeswitch_exporter-1.0.0-py3-none-any.whl.gz](https://github.com/znerol/prometheus-freeswitch-exporter/files/6265909/prometheus_freeswitch_exporter-1.0.0-py3-none-any.whl.gz)

**Comment by tyl0r** [comment]
I'm getting the same output

```
Exception thrown while rendering view
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/freeswitch_exporter/http.py", line 103, in view
    return self._views[endpoint](**params)
  File "/usr/local/lib/python3.8/dist-packages/freeswitch_exporter/http.py", line 55, in on_esl
    output = collect_esl(self._config[module], target)
  File "/usr/local/lib/python3.8/dist-packages/freeswitch_exporter/collector.py", line 301, in collect_esl
    return generate_latest(registry)
  File "/usr/local/lib/python3.8/dist-packages/prometheus_client/exposition.py", line 107, in generate_latest
    for metric in registry.collect():
  File "/usr/local/lib/python3.8/dist-packages/prometheus_client/registry.py", line 83, in collect
    for metric in collector.collect():
  File "/usr/local/lib/python3.8/dist-packages/asgiref/sync.py", line 147, in __call__
    return call_result.result()
  File "/usr/lib/python3.8/concurrent/futures/_base.py", line 432, in result
    return self.__get_result()
  File "/usr/lib/python3.8/concurrent/futures/_base.py", line 388, in __get_result
    raise self._exception
  File "/usr/local/lib/python3.8/dist-packages/asgiref/sync.py", line 212, in main_wrap
    result = await self.awaitable(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/freeswitch_exporter/collector.py", line 290, in collect
    await ESLChannelInfo(esl).collect())
  File "/usr/local/lib/python3.8/dist-packages/freeswitch_exporter/collector.py", line 239, in collect
    channelvars = json.loads(result)
  File "/usr/lib/python3.8/json/__init__.py", line 357, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3.8/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.8/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
127.0.0.1 - - [06/Apr/2021 09:05:33] "GET /esl?module=default&target=fs1.internal HTTP/1.1" 500 -
```

**Comment by znerol** [comment]
Ok, thanks for testing that. It looks like FS might be responding with `-ERR No such channel!` and that's obviously neither an empty string nor is it valid json. Thus it is necessary to cover that case as well. Pushed a fix into the other PR. A wheel including that fix is attached.

[prometheus_freeswitch_exporter-1.0.0-py3-none-any.whl.gz](https://github.com/znerol/prometheus-freeswitch-exporter/files/6266216/prometheus_freeswitch_exporter-1.0.0-py3-none-any.whl.gz)

**Comment by tyl0r** [comment]
Great, works now... though I had to change `startsWith` to `startswith` on line 235 in collector.py

**Comment by znerol** [comment]
> Great, works now... though I had to change startsWith to startswith on line 235 in collector.py

Oh, right. The other one is JavaScript syntax :sweat_smile: 

I removed the initial/ineffective fixes in order to cleanup the PR. It would be much appreciated if you could give that a (hopefully) final test-round.

[prometheus_freeswitch_exporter-1.0.0-py3-none-any.whl.gz](https://github.com/znerol/prometheus-freeswitch-exporter/files/6266968/prometheus_freeswitch_exporter-1.0.0-py3-none-any.whl.gz)

**Comment by tyl0r** [comment]
Yep, still looks good. 👍

**Comment by znerol** [comment]
Thanks for your help troubleshooting this issue. Released 1.0.1 on pypi which includes this fix.
